File: HanByeolKo/week018/day0121_Implementation/Bj20006.java
package week018.day0121_Implementation;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.StringTokenizer;

public class Bj20006 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int p = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());

        List<Room> rooms = new ArrayList<>();

        for (int i = 0; i < p; i++){
            st = new StringTokenizer(br.readLine());
            int level = Integer.parseInt(st.nextToken());
            String nickname = st.nextToken();
            Player player = new Player(nickname, level);

            boolean added = false;

            for (Room room : rooms){
                if(room.canJoin(player)){
                    room.addPlayer(player);
                    added = true;
                    break;
                }
            }

            if(!added){
                Room newRoom = new Room(m, level);
                newRoom.addPlayer(player);
                rooms.add(newRoom);
            }
        }

        for(Room room : rooms){
            if(room.isFull()){
                System.out.println("Started!");
            } else{
                System.out.println("Waiting!");
            }
            room.sortPlayers();
            for(Player player : room.players){
                System.out.println(player.level + " " + player.nickname);
            }
        }
    }

    static class Player{
        String nickname;
        int level;

        public Player(String nickname, int level){
            this.nickname = nickname;
            this.level = level;
        }
    }

    static class Room{
        int limit;
        int maxLevel;
        int minLevel;
        List<Player> players;

        public Room(int limit, int level){
            this.limit = limit;
            this.maxLevel = level + 10;
            this.minLevel = level - 10;
            this.players = new ArrayList<>();
        }

        public boolean canJoin(Player player){
            return players.size() < limit && player.level >= minLevel && player.level <= maxLevel;
        }

        public void addPlayer(Player player){
            players.add(player);
        }

        public void sortPlayers(){
            players.sort(Comparator.comparing(p->p.nickname));
        }

        public boolean isFull(){
            return players.size() == limit;
        }
    }
}

File: JisooPark/Week011_020/Week018/0121_Implementation/bj20006.cpp
#include <iostream> 
#include <vector>
#include <string> 
#include <algorithm>

using namespace std; 

int p, m;

struct Player 
{
    int level;
    string id;
};

struct Room 
{
    int minLevel;
    int maxLevel;

    vector<Player> playerList;
};


int main() 
{
    cin >> p >> m;

    vector<Room> rooms;     

    for(int idx = 0; idx < p; ++idx)
    {
        int level;
        string id; 
        cin >> level >> id;

        bool isAssigned = false;

        for(auto& room : rooms)
        {
            if(room.minLevel <= level && level <= room.maxLevel 
            && room.playerList.size() < m)
            {
                room.playerList.push_back({level, id});
                isAssigned = true;
                break;
            }
        }

        if(!isAssigned)
        {
            rooms.push_back({level - 10, level + 10, {{level, id}}});
        }
    }

    for(auto& room : rooms)
    {
        sort(room.playerList.begin(), room.playerList.end(), 
        [](const Player& a, const Player& b)
        {
            return a.id < b.id;
        });
        
        if(room.playerList.size() == m)
        {
            cout << "Started!\n";
        }
        else
        {
            cout << "Waiting!\n";
        }
        
        for(auto& player : room.playerList)
        {
            cout << player.level << ' ' << player.id << '\n';
        }
    }

    return 0;
}